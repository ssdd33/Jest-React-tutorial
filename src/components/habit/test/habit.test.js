import React from "react";
import Habit from "../habit";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
/*TODO:
1.props로 전달받은 habit이 화면에 렌더 되는지
2.버튼마다 event가 발생했을때 props로 전달받은 함수를 실행시키는지
 */

describe("habit", () => {
  let habitComponent;
  let onIncrement;
  let onDecrement;
  let onDelete;
  const habit = { name: "habit 1", count: 0 };

  beforeEach(() => {
    onIncrement = jest.fn();
    onDecrement = jest.fn();
    onDelete = jest.fn();
    habitComponent = (
      <Habit
        habit={habit}
        onIncrement={onIncrement}
        onDecrement={onDecrement}
        onDelete={onDelete}
      />
    );
  });
  it("renders", () => {
    const component = renderer.create(habitComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("button click", () => {
    beforeEach(() => {
      render(habitComponent);
    });
    it('calls onIncrement when clicking "increment" button', () => {
      const button = screen.getByTitle("increment");
      userEvent.click(button);
      expect(onIncrement).toHaveBeenCalledWith(habit);
    });
    it('calls onDecrement when clicking "decrement"button', () => {
      const button = screen.getByTitle("decrement");
      userEvent.click(button);
      expect(onDecrement).toHaveBeenCalledWith(habit);
    });
    it('calls onDelete when clicking "delete"button', () => {
      const button = screen.getByTitle("delete");
      userEvent.click(button);
      expect(onDelete).toHaveBeenCalledWith(habit);
    });
  });
});
