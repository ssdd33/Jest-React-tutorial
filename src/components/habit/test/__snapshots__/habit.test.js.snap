// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`habit renders 1`] = `
<li
  className="habit"
>
  <span
    className="habit-name"
  >
    habit 1
  </span>
  <span
    className="habit-count"
  >
    0
  </span>
  <button
    className="habit-button habit-increase"
    onClick={[Function]}
    title="increment"
  >
    <i
      className="fas fa-plus-square"
    />
  </button>
  <button
    className="habit-button habit-decrease"
    onClick={[Function]}
    title="decrement"
  >
    <i
      className="fas fa-minus-square"
    />
  </button>
  <button
    className="habit-button habit-delete"
    onClick={[Function]}
    title="delete"
  >
    <i
      className="fas fa-trash"
    />
  </button>
</li>
`;
